File: examples/human/cli/python/agent.py
"""Deploy Readiness Checker — human/cli example.

Delivery: stream  (GET /v1/agents/{address}/intents/stream)

Checks deployment readiness before a human operator makes the final call.
Validates: semver version format, allowed environment, non-empty rollback tag.

Run (Terminal 1 — agent):
    export AXME_API_KEY=<your-api-key>
    export AXME_AGENT_ADDRESS=agent://<org>/<workspace>/deploy-readiness-checker
    python agent.py

Run (Terminal 2 — scenario):
    axme scenarios apply examples/human/cli/scenario.json --watch

After the agent completes its step, the workflow pauses for human approval:
    axme tasks list
    axme tasks approve <task_id>   # or: axme tasks reject <task_id>
"""
from __future__ import annotations

import logging
import os
import re
import sys
from typing import Any

from axme import AxmeClient, AxmeClientConfig

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s  [%(levelname)s]  %(message)s",
    datefmt="%H:%M:%S",
)
log = logging.getLogger("deploy-readiness-checker")

AXME_BASE_URL      = os.environ.get("AXME_BASE_URL", "https://api.cloud.axme.ai")
AXME_API_KEY       = os.environ.get("AXME_API_KEY", "")
AXME_AGENT_ADDRESS = os.environ.get("AXME_AGENT_ADDRESS", "deploy-readiness-checker")

if not AXME_API_KEY:
    sys.exit("AXME_API_KEY is required")

# ---------------------------------------------------------------------------
# Semantic version pattern  (strict: MAJOR.MINOR.PATCH, no pre-release suffix)
# ---------------------------------------------------------------------------
_SEMVER_RE = re.compile(r"^\d+\.\d+\.\d+$")

_ALLOWED_ENVIRONMENTS = {"staging", "canary", "preview"}


def _check_readiness(payload: dict[str, Any]) -> tuple[bool, list[str], list[str]]:
    """Return (ready, passed_checks, failures)."""
    deploy_id   = payload.get("deploy_id", "?")
    version     = payload.get("version", "")
    environment = payload.get("environment", "")
    rollback    = payload.get("rollback_tag", "")
    service     = payload.get("service", "")

    failures: list[str] = []
    passed:   list[str] = []

    # Check 1 — version is semver
    if _SEMVER_RE.match(version):
        passed.append(f"version {version!r} is valid semver")
    else:
        failures.append(f"version {version!r} is not valid semver (expected MAJOR.MINOR.PATCH)")

    # Check 2 — environment is allowed for automated deployment
    if environment in _ALLOWED_ENVIRONMENTS:
        passed.append(f"environment {environment!r} is in allowed list")
    else:
        failures.append(
            f"environment {environment!r} not in allowed set: {sorted(_ALLOWED_ENVIRONMENTS)}"
        )

    # Check 3 — rollback tag specified
    if rollback:
        passed.append(f"rollback_tag {rollback!r} is set")
    else:
        failures.append("rollback_tag is missing — deployment must have a rollback path")

    # Check 4 — service name not empty
    if service:
        passed.append(f"service {service!r} is identified")
    else:
        failures.append("service name is empty")

    return len(failures) == 0, passed, failures


def handle_intent(client: AxmeClient, delivery: dict[str, Any]) -> None:
    intent_id = delivery.get("intent_id", "")
    if not intent_id:
        log.warning("received delivery without intent_id, skipping")
        return

    log.info("received intent %s", intent_id)

    try:
        resp   = client.get_intent(intent_id)
        intent = resp.get("intent") or resp
    except Exception as exc:
        log.error("get_intent(%s) failed: %s", intent_id, exc)
        return

    status = (intent.get("lifecycle_status") or intent.get("status") or "").upper()
    if status not in {"CREATED", "DELIVERED", "ACKNOWLEDGED", "IN_PROGRESS", "WAITING"}:
        log.debug("skipping intent %s with status %s", intent_id, status)
        return

    raw_payload      = intent.get("payload") or {}
    effective_payload = raw_payload.get("parent_payload") or raw_payload

    ready, passed, failures = _check_readiness(effective_payload)
    log.info(
        "readiness result for %s: ready=%s  passed=%d  failed=%d",
        intent_id, ready, len(passed), len(failures),
    )

    try:
        if ready:
            client.resume_intent(
                intent_id,
                {
                    "action":         "complete",
                    "ready":          True,
                    "passed_checks":  passed,
                    "failures":       [],
                },
                owner_agent=AXME_AGENT_ADDRESS,
            )
        else:
            client.resume_intent(
                intent_id,
                {
                    "action":        "fail",
                    "ready":         False,
                    "passed_checks": passed,
                    "failures":      failures,
                },
                owner_agent=AXME_AGENT_ADDRESS,
            )
        log.info("resumed intent %s (ready=%s)", intent_id, ready)
    except Exception as exc:
        log.error("resume_intent(%s) failed: %s", intent_id, exc)


def main() -> None:
    client = AxmeClient(AxmeClientConfig(api_key=AXME_API_KEY, base_url=AXME_BASE_URL))
    log.info(
        "deploy-readiness-checker starting  address=%s  binding=stream",
        AXME_AGENT_ADDRESS,
    )

    try:
        for delivery in client.listen(AXME_AGENT_ADDRESS, timeout_seconds=None):
            handle_intent(client, delivery)
    except KeyboardInterrupt:
        log.info("shutting down")


if __name__ == "__main__":
    main()

File: examples/human/cli/typescript/agent.ts
/**
 * Deploy Readiness Checker — human/cli example (TypeScript).
 *
 * Validates deployment readiness before routing to a human operator.
 * After the agent step, the workflow pauses for human approval via CLI.
 *
 * Run:
 *   export AXME_API_KEY=<your-api-key>
 *   export AXME_AGENT_ADDRESS=deploy-readiness-checker
 *   npx tsx examples/human/cli/agent.ts
 */
import { AxmeClient } from "@axme/axme";

const AXME_BASE_URL = process.env.AXME_BASE_URL ?? "https://api.cloud.axme.ai";
const AXME_API_KEY = process.env.AXME_API_KEY ?? "";
const AXME_AGENT_ADDRESS = process.env.AXME_AGENT_ADDRESS ?? "deploy-readiness-checker";

if (!AXME_API_KEY) { console.error("AXME_API_KEY is required"); process.exit(1); }

// --- Business logic ---
const SEMVER_RE = /^\d+\.\d+\.\d+$/;
const ALLOWED_ENVIRONMENTS = new Set(["staging", "canary", "preview"]);

function checkReadiness(payload: Record<string, unknown>): { ready: boolean; passed: string[]; failures: string[] } {
  const version = String(payload.version ?? "");
  const environment = String(payload.environment ?? "");
  const rollback = String(payload.rollback_tag ?? "");
  const service = String(payload.service ?? "");
  const passed: string[] = [];
  const failures: string[] = [];

  if (SEMVER_RE.test(version)) passed.push(`version '${version}' is valid semver`);
  else failures.push(`version '${version}' is not valid semver (expected MAJOR.MINOR.PATCH)`);

  if (ALLOWED_ENVIRONMENTS.has(environment)) passed.push(`environment '${environment}' is in allowed list`);
  else failures.push(`environment '${environment}' not in allowed set: ${[...ALLOWED_ENVIRONMENTS].sort()}`);

  if (rollback) passed.push(`rollback_tag '${rollback}' is set`);
  else failures.push("rollback_tag is missing — deployment must have a rollback path");

  if (service) passed.push(`service '${service}' is identified`);
  else failures.push("service name is empty");

  return { ready: failures.length === 0, passed, failures };
}

// --- Agent loop ---
const client = new AxmeClient({ apiKey: AXME_API_KEY, baseUrl: AXME_BASE_URL });
console.log(`deploy-readiness-checker starting  address=${AXME_AGENT_ADDRESS}  binding=stream`);

try {
  for await (const delivery of client.listen(AXME_AGENT_ADDRESS)) {
    const intentId = String(delivery.intent_id ?? "");
    if (!intentId) continue;
    console.log(`received intent ${intentId}`);

    let intent: Record<string, unknown>;
    try {
      const resp = await client.getIntent(intentId);
      intent = (resp.intent ?? resp) as Record<string, unknown>;
    } catch (e) { console.error(`get_intent(${intentId}) failed:`, e); continue; }

    const status = String(intent.lifecycle_status ?? intent.status ?? "").toUpperCase();
    if (!["CREATED", "DELIVERED", "ACKNOWLEDGED", "IN_PROGRESS", "WAITING"].includes(status)) continue;

    const rawPayload = (intent.payload ?? {}) as Record<string, unknown>;
    const effectivePayload = (rawPayload.parent_payload ?? rawPayload) as Record<string, unknown>;
    const { ready, passed, failures } = checkReadiness(effectivePayload);

    console.log(`readiness for ${intentId}: ready=${ready}  passed=${passed.length}  failed=${failures.length}`);

    try {
      if (ready) {
        await client.resumeIntent(intentId, { action: "complete", ready: true, passed_checks: passed, failures: [] }, { ownerAgent: AXME_AGENT_ADDRESS });
      } else {
        await client.resumeIntent(intentId, { action: "fail", ready: false, passed_checks: passed, failures }, { ownerAgent: AXME_AGENT_ADDRESS });
      }
      console.log(`resumed intent ${intentId} (ready=${ready})`);
    } catch (e) { console.error(`resume_intent(${intentId}) failed:`, e); }
  }
} catch (e) {
  if ((e as Error).name !== "AbortError") throw e;
}
